src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroProtectedIndexAccessEvaluator.java

```diff
@@ -0,0 +1,88 @@
+package com.amazon.opendistroforelasticsearch.security.privileges;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import com.amazon.opendistroforelasticsearch.security.auditlog.AuditLog;
+import com.amazon.opendistroforelasticsearch.security.resolver.IndexResolverReplacer;
+import com.amazon.opendistroforelasticsearch.security.securityconf.SecurityRoles;
+import com.amazon.opendistroforelasticsearch.security.support.ConfigConstants;
+import com.amazon.opendistroforelasticsearch.security.support.WildcardMatcher;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.RealtimeRequest;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.tasks.Task;
+
+public class OpenDistroProtectedIndexAccessEvaluator {
+
+    protected final Logger log = LogManager.getLogger(this.getClass());
+
+    private final AuditLog auditLog;
+    private final Collection<String> indexPatterns;
+    private final Collection<String> allowedRoles;
+    private final String[] deniedActionPatterns;
+
+
+    public OpenDistroProtectedIndexAccessEvaluator(final Settings settings, AuditLog auditLog) {
+        this.indexPatterns = settings.getAsList(ConfigConstants.OPENDISTRO_SECURITY_PROTECTED_INDICES_KEY, ConfigConstants.OPENDISTRO_SECURITY_PROTECTED_INDICES_DEFAULT);
+        this.allowedRoles = settings.getAsList(ConfigConstants.OPENDISTRO_SECURITY_PROTECTED_INDICES_ROLES_KEY, ConfigConstants.OPENDISTRO_SECURITY_PROTECTED_INDICES_ROLES_DEFAULT);
+        this.auditLog = auditLog;
+
+        final List<String> indexDeniedActionPatterns = new ArrayList<String>();
+        indexDeniedActionPatterns.add("indices:data/write*");
+        indexDeniedActionPatterns.add("indices:admin/delete*");
```

Comment:
Changed to `indices:admin/*` but need to list `indices:data/*` due to otherwise search failing with security exception.
Added snapshot operations to deny list as well: `"cluster:admin/snapshot/*`